src/deroff.rs

```diff
@@ -538,12 +538,311 @@ impl Deroffer {
+        }
+
+        self.s = self.skip_leading_whitespace(&self.s).to_owned();
+
+        loop {
+            if !self.quoted_arg() && !self.text_arg() {
+                if !self.s.is_empty() {
+                    // self.condputs(self.str_at(0));
+                    self.condputs(Self::str_at(&self.s, 0));
+                    self.s = self.skip_char(&self.s, None).to_owned();
+                } else {
+                    return true;
+                }
+            }
+        }
+    }
+
+    fn font(&mut self) -> bool {
+        if let Some(m) = self.g_re_font.find(&self.s) {
+            self.s = self.skip_char(&self.s, Some(m.end())).to_owned();
+            true
+        } else {
+            false
+        }
+    }
+
+    fn comment(&self) -> bool {
+        unimplemented!()
+    }
+
+    fn numreq(&mut self) -> bool {
+        // In the python, it has a check that is already handled in esc_char_backslash, which is
+        // the only place it gets called, so I'll omit that check here
+
+        // This is written as `self.macro += 1` in the source, but I dont know why
+        // it does the same thing (false -> true, true -> still true) :shrug:
+        // self.r#macro = true;
+        // Upon further investigation, this is the weirdest function ever
+        // This is just a state placeholder thing
+        let mut m = self.r#macro as u8 + 1;
+
+        // self.skip_char(3);
+        self.s = self.skip_char(&self.s, Some(3)).to_owned();
+
+        // There's this, which has a comment explaining my thoughts right now
+        /*
+        while self.str_at(0) != "'" and self.esc_char():
+            pass  # Weird
+        */
+        // And I dont know what the purpose of it would even be, if you can tell, lmk
+
+        // if self.str_at(0) == '\'' {
+        if Self::str_at(&self.s, 0) == "'" {
+            // self.skip_char(1);
+            self.s = self.skip_char(&self.s, None).to_owned();
+        }
+
+        m -= 1;
+        self.r#macro = m != 0;
+
+        true
+    }
+
+    fn text_arg(&self) -> bool {
+        unimplemented!()
+    }
+
+    fn prch(&self, idx: usize) -> bool {
+        self.s
+            .get(idx..=idx)
+            .map(|c| !" \t\n".contains(c))
+            .unwrap_or_default()
+    }
+
+    // This function is the worst, there are a few comments explaining some of it in the test (test_var)
+    // its so hard to briefly put into words what this function does, basically depending on the state
+    // of self.s, it will either, change self.s to "", a part of self.s, or a value in self.reg_table
+    // which corresponds to a key that is part of self.s.
+    // This should be like 2 or 3 functions, but it's only one. So there's that. :-)
+    // NOTE: there is a call to text_arg that is commented out because it's not implemented, so the
+    // tests will need revised when it gets implemented
+    fn var(&mut self) -> bool {
+        let s0s1 = &self.s[0..2];
```

Comment:
This may fail, it should use something safer.

Reply:
Fixed:
```rust
let s0s1 = match self.s.get(0..2) {
    Some(s) => s,
    None => return false,
};
```